src/update_agent/actor.rs

```diff
@@ -381,27 +394,39 @@ impl UpdateAgent {
         release: Release,
     ) -> ResponseActFuture<Self, Result<Release, ()>> {
         if !can_finalize {
+            // Reset number of postponements if finalization attempt failed due to
```

Comment:
There are at least two state transitions nested very deeply inside this function, and I think it would better to split the concerns so that in the top-level `tick_finalize_update` it is clearly visible that there three cases:
 * `strategy_can_finalize: false` -> reset the number of attempts
 * `strategy_can_finalize: true` -> handle interactive users, then:
   * `usersessions_can_finalize: false` -> update the number of attempts
   * `usersessions_can_finalize: true` -> call `finalize_deployment(true)`

Reply:
I've tried to address this in the latest WIP commit. I've made it a WIP commit because I feel like I actually made the code more difficult to read. It'd be great if you can look at the diff of just that commit and see if I'm going in the right general direction. Thanks!